### PR TITLE
ELK v8.x analytic configurations

### PR DESCRIPTION
**Pull request by Avarjana**
## Description 
Added required configurations for ELK analytics support.

### CODE REVIEW

**Review by IsankaSR — APPROVED** [review]

File: Kibana/transforms/transform.curl [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    "buckets_path": {
+                        "failed": "filter_failed._count",
+                        "success": "filter_success._count"
+                    },
+                    "script": "if(params.success >= 1 && params.failed >= 2 && params.failed > params.success) return 1; else return 0;"
+                }
+            }
+        }
+    }
+}'
```
Comment (bhagyasakalanka) [inline Kibana/transforms/transform.curl:86]: shall we add a new line?

File: Logstash/logstash_filebeats.conf [hunk 2]
```diff
@@ -1,73 +1,137 @@
...
 	ruby{
-	     code =>"
+		code =>"
+
+		# method to split the supplied string by comma, trim whitespace and return an array
+		def mapStringToArray(strFieldValue)
 
-		 # method to split the supplied string by comma, trim whitespace and return an array
-		 def mapStringToArray(strFieldValue)
+		#if string is not null, return array
```
Comment (bhagyasakalanka) [inline Logstash/logstash_filebeats.conf:39]: Lets follow standard when commenting. Start with a capital letter and end with a full stop. Add a space between # and first letter.  `# If string is not null, return array.`

Look for other places as well

File: Logstash/logstash_filebeats.conf [hunk 3]
```diff
@@ -1,73 +1,137 @@
...
-	geoip {
-    		source => "[event][payloadData][remoteIp]"
-    	}
 
 	#ruby code to convert the comma seperated roles into an array
 	ruby{
-	     code =>"
+		code =>"
+
+		# method to split the supplied string by comma, trim whitespace and return an array
```
Comment (bhagyasakalanka) [inline Logstash/logstash_filebeats.conf:36]: commenting issue

File: Logstash/logstash_filebeats.conf [hunk 4]
```diff
@@ -1,73 +1,137 @@
...
 	}
+
 }
 
 output {
+
 	#seperating event type as auth and session via the eventtype
-	if [eventType] == "auth" {
+
+	if [eventType] == "auth"{
```
Comment (bhagyasakalanka) [inline Logstash/logstash_filebeats.conf:111]: Shall we add a space bethween "auth" and { 
`"auth" {`. Look for other places as well

File: Logstash/logstash_filebeats.conf [hunk 5]
```diff
@@ -1,73 +1,137 @@
...
-			index => "auth"
+			hosts => ["https://localhost:9200"]
+			cacert => "/Users/avarjana/Documents/Work/ELK/logstash-8.1.2/config/certs/http_ca.crt"
+			user => "<Elastic Username>"
+			password => "<Elastic User Password>"
+			index => "auth-events"
+			document_id => "%{[event][payloadData][eventId]}"
 		}
-	} else if [eventType] == "session" {
+	} else if [eventType] == "session"{
```
Comment (bhagyasakalanka) [inline Logstash/logstash_filebeats.conf:120]: formatting issue

File: Logstash/logstash_http.conf [hunk 6]
```diff
@@ -19,52 +24,100 @@ filter {
...
-		 def mapStringToArray(strFieldValue)
+		# method to split the supplied string by comma, trim whitespace and return an array
+		def mapStringToArray(strFieldValue)
 
-		    #if string is not null, return array
-		    if (strFieldValue != nil)
-		        fieldArr =  strFieldValue.split(',').map(&:strip).reject(&:empty?).uniq 
-		        return fieldArr                             
-		    end     
+		#if string is not null, return array
```
Comment (bhagyasakalanka) [inline Logstash/logstash_http.conf:32]: issues in comment line. look for other places as well

File: Logstash/logstash_http.conf [hunk 7]
```diff
@@ -19,52 +24,100 @@ filter {
...
+		if (strFieldValue != nil)
+			fieldArr =  strFieldValue.split(',').map(&:strip).reject(&:empty?).uniq 
+			return fieldArr                             
+		end     
 
-		    return [] #return empty array if string is nil
-		 end
+		return [] #return empty array if string is nil
```
Comment (bhagyasakalanka) [inline Logstash/logstash_http.conf:38]: issues in comments

File: Logstash/logstash_http.conf [hunk 8]
```diff
@@ -19,52 +24,100 @@ filter {
...
+		vrtArr = mapStringToArray(event.get('[event][payloadData][rolesCommaSeparated]'))
+		if vrtArr.length > 0                           
+		event.set('[event][payloadData][rolesCommaSeparated]', vrtArr) 
+		end
+
+		event_array = []
+		event_type = event.get('[event][payloadData][eventType]')
+		idp_type = event.get('[event][payloadData][identityProviderType]')
+		auth_step_success = event.get('[event][payloadData][authStepSuccess]')
+        if (event_type == 'overall' || auth_step_success == false)
```
Comment (bhagyasakalanka) [inline Logstash/logstash_http.conf:50]: wrong indentations

File: Logstash/logstash_http.conf [hunk 9]
```diff
@@ -19,52 +24,100 @@ filter {
...
+		# user agent filter to transform the userAgent string into seperated fields
+		useragent {
+			ecs_compatibility => disabled
+			source => "[event][payloadData][userAgent]"
+			target => "userAgentDetails"
+		}
+
+	} else if [eventType] == "auth" {
+
+		#geo ip filter to transform the IP address to location
```
Comment (bhagyasakalanka) [inline Logstash/logstash_http.conf:77]: fix the comment format

File: Logstash/logstash_http.conf [hunk 10]
```diff
@@ -19,52 +24,100 @@ filter {
...
+		
 	}
 }
 
 output {
+
 	#seperating event type as auth and session via the eventtype header
-	if [headers][eventtype] == "auth"{
+
+	if [@metadata][input][http][request][headers][eventtype] == "auth"{
```
Comment (bhagyasakalanka) [inline Logstash/logstash_http.conf:97]: formatting issue

File: sso-oidc/elasticsearch-with-oidc.yml [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+# Allow HTTP API connections from localhost and local networks
+# Connections are encrypted and require user authentication
+http.host: [_local_, _site_]
+
+# Allow other nodes to join the cluster from localhost and local networks
+# Connections are encrypted and mutually authenticated
+#transport.host: [_local_, _site_]
+
+#----------------------- END SECURITY AUTO CONFIGURATION -------------------------
```
Comment (bhagyasakalanka) [inline sso-oidc/elasticsearch-with-oidc.yml:139]: shall we add a new line

File: sso-oidc/kibana-with-oidc.yml [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+xpack.fleet.outputs: [{id: fleet-default-output, name: default, is_default: true, is_default_monitoring: true, type: elasticsearch, hosts: ['https://192.168.1.3:9200'], ca_trusted_fingerprint: f577b7da7dd4817616b5fe7ce293ba5a8883d1ea4e8e56ab2dc616e4ea3f048d}]
+
+# SSO Related Configs 
+xpack.security.authc.providers:
+  basic.basic1:
+    order: 0
+  oidc.oidc1:
+    order: 1
+    realm: "oidc1"
+    description: "Log in with WSO2 via OIDC"
```
Comment (bhagyasakalanka) [inline sso-oidc/kibana-with-oidc.yml:48]: shall we add a newline

File: FileBeats/fb.yml [hunk 13]
```diff
@@ -1,243 +1,16 @@
...
-  # Protocol - either `http` (default) or `https`.
-  #protocol: "https"
-
-  # Authentication credentials - either API key or username/password.
-  #api_key: "id:api_key"
-  #username: "elastic"
-  #password: "changeme"
-
-# ------------------------------ Logstash Output -------------------------------
+  paths:
```
Comment (lashinijay) [inline FileBeats/fb.yml:11]: formatting issue. is the space in the beginning intentionally added?

File: Kibana/saved-objects/kibana-8-x-auth-and-session.ndjson [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+{"excludedObjects":[],"excludedObjectsCount":0,"exportedCount":33,"missingRefCount":0,"missingReferences":[]}
```
Comment (lashinijay) [inline Kibana/saved-objects/kibana-8-x-auth-and-session.ndjson:34]: shall we add a new line here? check other places also.

File: Logstash/logstash_filebeats.conf [hunk 15]
```diff
@@ -1,73 +1,137 @@
...
 	}
-	#geo ip filter to transform the IP address to location
-	geoip {
-    		source => "[event][payloadData][remoteIp]"
-    	}
 
 	#ruby code to convert the comma seperated roles into an array
 	ruby{
-	     code =>"
+		code =>"
```
Comment (lashinijay) [inline Logstash/logstash_filebeats.conf:34]: formatting issue

File: Logstash/logstash_filebeats.conf [hunk 16]
```diff
@@ -1,73 +1,137 @@
...
 
-		 vrtArr = mapStringToArray(event.get('[event][payloadData][rolesCommaSeparated]'))
-		 if vrtArr.length > 0                           
-		    event.set('[event][payloadData][rolesCommaSeparated]', vrtArr) 
-		 end
+		event_array = []
+		event_type = event.get('[event][payloadData][eventType]')
+		idp_type = event.get('[event][payloadData][identityProviderType]')
+		auth_step_success = event.get('[event][payloadData][authStepSuccess]')
+        if (event_type == 'overall' || auth_step_success == false)
```
Comment (lashinijay) [inline Logstash/logstash_filebeats.conf:57]: indentation issue

File: Logstash/logstash_http.conf [hunk 17]
```diff
@@ -1,15 +1,20 @@
 input {
 	http {
-		host => "127.0.0.1" # default: 0.0.0.0
-		port => 8084 # default: 8080
+		host => "0.0.0.0" # default: 0.0.0.0
+		port => 8081 # default: 8080
 	}
 }
 
+## Add your filters / logstash plugins configuration here
```
Comment (lashinijay) [inline Logstash/logstash_http.conf:8]: Add a full stop at the end. check other places also.

File: Logstash/logstash_http.conf [hunk 18]
```diff
@@ -19,52 +24,100 @@ filter {
...
+			source => "[event][payloadData][remoteIp]"
+		}
+
+		# geo IP failure to lookup or local address 
+		if "_geoip_lookup_failure" in [tags] {
+			mutate {
+				add_field => { "[geoip][country_name]" => "N/A" }
+			}
 		}
+
```
Comment (lashinijay) [inline Logstash/logstash_http.conf:89]: Let's remove the unnecessary line

File: Logstash/logstash_http.conf [hunk 19]
```diff
@@ -19,52 +24,100 @@ filter {
...
-			index => "auth"
+			hosts => ["https://localhost:9200"]
+			cacert => "/Users/avarjana/Documents/Work/ELK/logstash-8.1.2/config/certs/http_ca.crt"
+			user => "<Elastic Username>"
+			password => "<Elastic User Password>"
+			index => "auth-events"
+			document_id => "%{[event][payloadData][eventId]}"
 		}
-	} else if [headers][eventtype] == "session"{
+	} else if [@metadata][input][http][request][headers][eventtype] == "session"{
```
Comment (lashinijay) [inline Logstash/logstash_http.conf:106]: formatting issue in `"session"{`. Let's add a space between " and {

File: sso-oidc/elasticsearch-with-oidc.yml [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+#
+# Path to log files:
+#
+# path.logs: /var/log/elasticsearch
+#
+# ----------------------------------- Memory -----------------------------------
+#
+# Lock the memory on startup:
+#
+#bootstrap.memory_lock: true
```
Comment (lashinijay) [inline sso-oidc/elasticsearch-with-oidc.yml:43]: formatting issue, check other places in the files also.

Comment (Avarjana) [reply sso-oidc/elasticsearch-with-oidc.yml:43]: Some of these code lines are auto generated from ELK and referred in their documentation. Therefore, I addressed similar issues in the customized code lines.

File: Kibana/transforms/transform.curl (hunk 1 above)
Comment (Avarjana) [reply Kibana/transforms/transform.curl:86]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_filebeats.conf (hunk 2 above)
Comment (Avarjana) [reply Logstash/logstash_filebeats.conf:39]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_filebeats.conf (hunk 3 above)
Comment (Avarjana) [reply Logstash/logstash_filebeats.conf:36]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_filebeats.conf (hunk 4 above)
Comment (Avarjana) [reply Logstash/logstash_filebeats.conf:111]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_filebeats.conf (hunk 5 above)
Comment (Avarjana) [reply Logstash/logstash_filebeats.conf:120]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_http.conf (hunk 6 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:32]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_http.conf (hunk 7 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:38]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_http.conf (hunk 8 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:50]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_http.conf (hunk 9 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:77]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: Logstash/logstash_http.conf (hunk 10 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:97]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: sso-oidc/elasticsearch-with-oidc.yml (hunk 11 above)
Comment (Avarjana) [reply sso-oidc/elasticsearch-with-oidc.yml:139]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: sso-oidc/kibana-with-oidc.yml (hunk 12 above)
Comment (Avarjana) [reply sso-oidc/kibana-with-oidc.yml:48]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/14

File: FileBeats/fb.yml (hunk 13 above)
Comment (Avarjana) [reply FileBeats/fb.yml:11]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

File: Kibana/saved-objects/kibana-8-x-auth-and-session.ndjson (hunk 14 above)
Comment (Avarjana) [reply Kibana/saved-objects/kibana-8-x-auth-and-session.ndjson:34]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

File: Logstash/logstash_filebeats.conf (hunk 15 above)
Comment (Avarjana) [reply Logstash/logstash_filebeats.conf:34]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

File: Logstash/logstash_filebeats.conf (hunk 16 above)
Comment (Avarjana) [reply Logstash/logstash_filebeats.conf:57]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

File: Logstash/logstash_http.conf (hunk 17 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:8]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

File: Logstash/logstash_http.conf (hunk 18 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:89]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

File: Logstash/logstash_http.conf (hunk 19 above)
Comment (Avarjana) [reply Logstash/logstash_http.conf:106]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

File: sso-oidc/elasticsearch-with-oidc.yml (hunk 20 above)
Comment (Avarjana) [reply sso-oidc/elasticsearch-with-oidc.yml:43]: Fixed with https://github.com/wso2-extensions/identity-elk-integration/pull/15

**Pull request merged**